codice/ddf
# DDF-5627 G-7443 Add an endpoint to validate different aspects of a query

#### What does this PR do?
Adds a new framework for validating queries **before they are ran**. The idea is that as a user creates/modifies a query in the UI, it will be sent to the backend for validation in real-time.

 It works by exposing a new `cql/validator/:id` POST endpoint that accepts a CQL query body just like the `/cql` endpoint. The `:id` parameter is the id of the validator that you want to run. The response contains a set of validation warnings or errors.

New validators are added by implementing the `QueryValidator` interface. Currently, the only validator in the system is the `UnsupportedAttributesQueryValidator` with id `unsupportedAttribute`. This validator aggregates all instances of the `SourceAttributeRestriction` interface and uses them to determine if there are any attributes in the query that are not supported by the currently selected sources.

For example, a request to `/cql/validator/unsupportedAttribute` with the following body
```
{
    "srcs": ["OS"],
    "start": 1,
    "count": 250,
    "cql": "(\"ext.not-supported-by-os\" ILIKE 'some value')",
    "sorts": [
        {
            "attribute": "modified",
            "direction": "descending"
        }
    ],
    "id": "a65ffb0790194aeca46f37c9b1f14ccf",
    "spellcheck": false,
    "phonetics": false,
    "batchId": "106e57b9cff2461aa08486cc52064296"
}
```
... might produce a response like below.

```
{
    "validationViolations": [
        {
            "type": "unsupportedAttribute",
            "severity": "ERROR",
            "message": "The field \"ext.not-supported-by-os\" is not supported by the OS Source(s)",
            "extraData": {
                "attribute": "ext.not-supported-by-os",
                "sources": [
                    "OS"
                ]
            }
        }
    ]
}
```


Each validation violation is required to have a `type` (the id of the validator), a `severity`, and a `message`. The `message` field is user-friendly so that they can be shown directly to the user in the query view in Intrigue. The `extraData` field is optional and is used to return structured data about the violation so that no additional parsing is necessary to get key information (in this case, we won't need to parse the message to determine the name of the unsupported attribute or the sources that don't support it - they are supplied directly).

The approach is extensible to support other types of validation in the future.

#### Who is reviewing it? 

@brianfelix @kentmorrissey @brhumphe @Lambeaux @mazarag2 
#### Select relevant component teams: 


#### Ask 2 committers to review/merge the PR and tag them here.
@bdeining
@millerw8

#### How should this be tested?

1. Install the standard profile
2. Create a new OpenSearch source (Catalog OpenSearch Federated Source) called `OS`
3. Download/extract this zip and import the json file into Postman
[Validation.postman_collection.json.zip](https://github.com/codice/ddf/files/4055684/Validation.postman_collection.json.zip)
4. Run the `CQL Validate` request and ensure no validation errors/warnings are shown (this is expected since there is no `SourceAttributeRestriction` implementation for the OpenSearch interface by default)

##### Test the `UnsupportedAttributesQueryValidator` with a `SourceAttributeRestriction` implementation for OpenSearch
1. Remove the `OS` source
2. Cherry-pick https://github.com/mojogitoverhere/ddf/commit/e97c92722e1c2cb787329eb295c722e80b3b5521, cd to catalog/opensearch/catalog-opensearch-source, and rebuild
3. Hot deploy the changes using `bundle:watch` or by dropping the catalog-opensearch-source-2.22.0-SNAPSHOT.jar into the deploy folder
4. Recreate the `OS` source
5. Run the `CQL Validate` request again and verify that no errors/warnings are returned still (since `title` is the only attribute supported by the OS source)
6. Change `title` in the `cql` field to another attribute like `subtitle` and rerun `CQL Validate`. Verify that an error like below is returned
```
{
    "validationViolations": [
        {
            "type": "unsupportedAttribute",
            "severity": "ERROR",
            "message": "The field \"ext.title\" is not supported by the OS Source(s)",
            "extraData": {
                "attribute": "ext.title",
                "sources": [
                    "OS"
                ]
            }
        }
    ]
}
```
7. Test different combinations of attributes (including operators like contains, equals, etc.) and cql queries. Instead of manually typing the cql string, I recommend running a query in Intrigue and copy/pasting the cql from the outgoing request in the network tab. You can also change the set of supported attributes here https://github.com/mojogitoverhere/ddf/commit/e97c92722e1c2cb787329eb295c722e80b3b5521#diff-1e3c7ae03e5895233d8e433f676aaac5R1169, rebuild, and redeploy for more testing.


#### What are the relevant tickets?
Fixes: #5627 

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
